tests/test_collection.py

```diff
@@ -9,101 +10,86 @@
-        del collection
-
-        # access geometry, this should not seg fault as 1.2.15 did
-        self.assertIsNotNone(child.wkt)
-
-    @shapely20_deprecated
-    def test_geointerface_adapter(self):
-        d = {"type": "GeometryCollection","geometries": [
-                {"type": "Point", "coordinates": (0, 3)},
-                {"type": "LineString", "coordinates": ((2, 0), (1, 0))}
-            ]}
-
-        # asShape
-        m = asShape(d)
-        self.assertEqual(m.geom_type, "GeometryCollection")
-        self.assertEqual(len(m), 2)
-        geom_types = [g.geom_type for g in m.geoms]
-        self.assertIn("Point", geom_types)
-        self.assertIn("LineString", geom_types)
-
-    def test_geointerface(self):
-        d = {"type": "GeometryCollection","geometries": [
-                {"type": "Point", "coordinates": (0, 3)},
-                {"type": "LineString", "coordinates": ((2, 0), (1, 0))}
-            ]}
-
-        # shape
-        m = shape(d)
-        self.assertEqual(m.geom_type, "GeometryCollection")
-        self.assertEqual(len(m), 2)
-        geom_types = [g.geom_type for g in m.geoms]
-        self.assertIn("Point", geom_types)
-        self.assertIn("LineString", geom_types)
-
-    @shapely20_deprecated
-    def test_empty_geointerface_adapter(self):
-        d = {"type": "GeometryCollection", "geometries": []}
-
-        # asShape
-        m = asShape(d)
-        self.assertEqual(m.geom_type, "GeometryCollection")
-        self.assertEqual(len(m), 0)
-        self.assertEqual(m.geoms, [])
-
-    def test_empty_geointerface(self):
-        d = {"type": "GeometryCollection", "geometries": []}
-
-        # shape
-        m = shape(d)
-        self.assertEqual(m.geom_type, "GeometryCollection")
-        self.assertEqual(len(m), 0)
-        self.assertEqual(m.geoms, [])
-
-    def test_empty_coordinates(self):
-
-        d = {"type": "GeometryCollection", "geometries": [
-            {"type": "Point", "coordinates": ()},
-            {"type": "LineString", "coordinates": (())}
-        ]}
-
-        # shape
-        m = shape(d)
-        self.assertEqual(m.geom_type, "GeometryCollection")
-        self.assertEqual(len(m), 0)
-        self.assertEqual(m.geoms, [])
-
-
-def test_geometrycollection_adapter_deprecated():
-    d = {"type": "GeometryCollection","geometries": [
-            {"type": "Point", "coordinates": (0, 3)},
-            {"type": "LineString", "coordinates": ((2, 0), (1, 0))}
+@pytest.fixture()
+def geometrycollection_geojson():
+    return {"type": "GeometryCollection", "geometries": [
+        {"type": "Point", "coordinates": (0, 3, 0)},
+        {"type": "LineString", "coordinates": ((2, 0), (1, 0))}
     ]}
+
+
+@pytest.mark.parametrize('geom', [
+    GeometryCollection(),
+    shape({"type": "GeometryCollection", "geometries": []}),
+    shape({"type": "GeometryCollection", "geometries": [
+        {"type": "Point", "coordinates": ()},
+        {"type": "LineString", "coordinates": (())}
```

Comment:
Should this be considered as empty? The subparts are empty geometries, but I would expect the actual GeometryCollection to have a length of 2? 
(or are the empty subparts skipped on construction?)

Reply:
(now, given this was in the original tests as well, that's certainly for a separate issue to discuss ;))

Reply:
Glad you brought this up! I thought it was a little funky when I wrote it. I'm happy the test was there because it demonstrated behavior I certainly would not have expected.

It feels like something we should address, but I think you're right - it should probably as a separate issue.

Reply:
Empty strikes again! JTS/GEOS/PostGIS regard this geometry to be empty. There are, however, subtle differences. GEOS and JTS appear to convert this simply to 'GEOMETRYCOLLECTION EMPTY', while PostGIS keeps the empty items for WKT, but doesn't count them as geometries, e.g.:
```
db=> select ST_IsEmpty(g), ST_NumGeometries(g), ST_AsText(g), ST_AsText(ST_GeometryN(g, 1)), ST_AsText(ST_GeometryN(g, 2))
db-> from (select 'GeometryCollection(Point empty, LineString empty)'::geometry g) f;
-[ RECORD 1 ]----+-------------------------------------------------
st_isempty       | t
st_numgeometries | 0
st_astext        | GEOMETRYCOLLECTION(POINT EMPTY,LINESTRING EMPTY)
st_astext        | POINT EMPTY
st_astext        | LINESTRING EMPTY
```
Why, however, ST_NumGeometries is zero but ST_GeometryN for 1 and 2 works is suspicious.

Reply:
Weird, thanks for the example. Consistency with JTS and GEOS in these details seems nice from the perspective of someone familiar with those tools, but I wonder where we draw the line. After all, I'd expect `[[], []]` in Python to be truth-y and have a length of two, and it feels sort of wrong to change that when we're talking about what is essentially a list of geometries.

Reply:
It seems that GEOS also keeps the empty geometries, at least with PyGEOS I get:

```
In [39]: g = pygeos.Geometry("GEOMETRYCOLLECTION( POINT EMPTY, POINT EMPTY )") 

In [40]: g   
Out[40]: <pygeos.Geometry GEOMETRYCOLLECTION EMPTY>

In [41]: pygeos.get_num_geometries(g)
Out[41]: 2
```

So it still indicates that the collection has 2 geometries, although the WKT version just says "EMPTY" (probably because the collection is regarded as empty if all subparts are empty):

```
In [42]: pygeos.is_empty(g)   
Out[42]: True
```

Reply:
There appears to be multiple bugs in this geometry sub-genre, such as [the PostGIS one mentioned above](https://trac.osgeo.org/postgis/ticket/4736), that [these geometries do not accurately serialize to WKT or WKB](https://trac.osgeo.org/geos/ticket/1040).

This is a minor shapely bug:
```python
from shapely import geos, wkt
g = wkt.loads("GEOMETRYCOLLECTION(LINESTRING EMPTY, POLYGON EMPTY)")
print(geos.lgeos.GEOSGetNumGeometries(g._geom))  # 2
print(len(g))  # 0
print(len(g.geoms))  # 0
```

Reply:
I moved this discussion to a new issue: https://github.com/Toblerity/Shapely/issues/956